lib/reboot_cmds.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python3
+
+"""
+Some uaclient operations cannot be fully completed by running a single
+command. For example, when upgrading uaclient from trusty to xenial,
+we may have a livepatch change in the contract, allowing livepatch to be
+enabled on xenial. However, during the upgrade we cannot install livepatch on
+the system, only after a reboot.
+
+To allow uaclient to postpone commands that need to be executed in a system boot,
+we are using this script, which will basically try to reprocess contract deltas
+on boot time.
+"""
+import logging
+import os
+import sys
+
+from uaclient import config
+
+from uaclient.util import subp, ProcessExecutionError
+from uaclient.cli import setup_logging, assert_lock_file
+
+
+CMDS_FILE_LOCATION = "/etc/ubuntu-advantage/reboot-cmds-needed"
+
+
+def run_command(cmd):
+    try:
+        out, _ = subp(cmd.split())
+        logging.debug("Successfully executed cmd: {}".format(cmd))
+    except ProcessExecutionError as exec_error:
+        msg = (
+            "Failed running cmd: {}\n"
+            "Return code: {}\n"
+            "Stderr: {}\n"
+            "Stdout: {}".format(
+                cmd, exec_error.exit_code, exec_error.stderr, exec_error.stdout
+            )
+        )
+        logging.debug(msg)
+        sys.exit(1)
+
+
+def refresh_contract(args, cfg):
+    cmd = "ua refresh"
+    run_command(cmd=cmd)
+
+
+@assert_lock_file("ua-reboot-process-deltas")
+def process_remaining_deltas(args, cfg):
+    cmd = "/usr/bin/python3 /usr/lib/ubuntu-advantage/upgrade_lts_contract.py"
```

Comment:
Is there somewhere that this is triggered based on facts of state? This is going to be run every startup. I understand refreshing with ua refresh, but I'm missing the step that interprets the changes in the state that determines that upgrade_lts_contract.py needs to run?

Reply:
The approach that we followed here is that the `upgrade_lts_contract` should be robust enough to identify if the system is already in the state that we desire. And if that is the case, it will not perform any operations. For example, suppose we have two deltas:

esm-infra: aptURL changes
livepathc: enabledByDefault is now True

If esm-infra is already configured with the aptURL delta and livepatch is already enabled, we will not perform anything.

Also, if the script succeeds on running all of the commands, we will no longer run it after every reboot.